### PR TITLE
Introduce _typeshed.Maybe

### PR DESCRIPTION
**Pull request by srittau**
We use the `X | Any` pattern for return types more and more. Introducing a type alias legitimizes the pattern, makes it easier to recognize, easier to grep, and means that we don't need to repeat the documentation.

### CODE REVIEW

File: stdlib/_typeshed/__init__.pyi [hunk 1]
```diff
@@ -33,6 +33,17 @@ Self = TypeVar("Self")  # noqa: Y001
 # "Incomplete | None" instead of "Any | None".
 Incomplete: TypeAlias = Any
 
+# Annotation for return types that can return a value or None, but where the
+# caller should not be forced to check for None. For example:
+#
+#   if m := re.match(r"(a)(b)?"):
```
Comment (Akuli) [inline stdlib/_typeshed/__init__.pyi:39]:
```suggestion
#   if m := re.match(r"(a)(b)?", string):
```

**Review by Akuli — COMMENTED** [review]
I like getting rid of the comment copy/pasta.

I'm not sure if `Maybe` is a good name for this, since it means "could be None and you must check for it" in some other languages, but I don't have better suggestions in mind.

**Comment by srittau** [comment]
> I'm not sure if `Maybe` is a good name for this, since it means "could be None and you must check for it" in some other languages, but I don't have better suggestions in mind.

True, but it's succinct and *has* the meaning "could be unset/null/none/undefined" in other languages, so the name at least has some meaning, even if it's not a 100% match to what other languages do.

**Comment by hauntsaninja** [comment]
We've talked about WeakUnion before, so WeakOptional seems a pretty natural name to me

**Comment by srittau** [comment]
I don't think we should use "optional". That name has caused lots of confusion in the past. Also, it's a fairly long name.

**Comment by Akuli** [comment]
`RarelyNone`?

**Comment by srittau** [comment]
We could bikeshed this forever, but I still think `Maybe` is the best option.

**Review by Akuli — APPROVED** [review]

**Comment by AlexWaygood** [comment]
> We could bikeshed this forever

Indeed we could, but I do think the choice of name is important here -- I'd prefer the status quo over a confusing alias name. Personally, I prefer either `Probably` or `RarelyNone` to `Maybe` 🙂

**Comment by srittau** [comment]
`Probably` and `RarelyNone` suggest that `None` is somehow the less common case. That's not true in many cases, for example `Match.group()` with optional groups.

**Comment by AlexWaygood** [comment]
`Often`? That has less of an implication that the result is necessarily the _majority_ of cases.

**Comment by Akuli** [comment]
I think the name shouldn't be too clever or short. If you see `Often` in the middle of some code, how are you supposed to know that it has something to do with unions and `None`, let alone guess its meaning?

Maybe we should just call this `TheAnyTrick`. It's very obvious for typeshed maintainers. It isn't obvious at first for people new to typing, but this is a difficult concept, and I don't think we can make it guessable anyway.

**Comment by AlexWaygood** [comment]
> Maybe we should just call this `TheAnyTrick`. It's very obvious for typeshed maintainers. It isn't obvious at first for people new to typing, but this is a difficult concept, and I don't think we can make it guessable anyway.

Or maybe just `AnyOr`? `Any | str` becomes `AnyOr[str]`

**Comment by AlexWaygood** [comment]
> Maybe we should just call this `TheAnyTrick`. It's very obvious for typeshed maintainers. It isn't obvious at first for people new to typing, but this is a difficult concept, and I don't think we can make it guessable anyway.

I'd be fine with `AnyTrick`. I agree it's better to have an opaque name than a name that's too clever or succinct.

**Comment by srittau** [comment]
No consensus.

**Pull request closed without merging**